cartoframes/data/services/geocode.py

```diff
@@ -0,0 +1,484 @@
+            geocoded_dataset, info = gc.geocode(dataframe, street='address', city='city', country="'Spain'")
+            print(dataframe)
+
+        Geocode a table:
+
+        .. code::
+
+            from data.services import Geocode
+            from cartoframes.data import Dataset
+            import pandas
+
+            dataframe = pandas.DataFrame([['Gran Vía 46', 'Madrid'], ['Ebro 1', 'Sevilla']], columns=['address','city'])
+            dataset = Dataset(dataframe)
+            dataset.upload(table_name='offices')
+            gc = Geocode()
+            geocoded_dataset, info = gc.geocode(dataset, street='address', city='city', country="'Spain'")
+            print(geocoded_dataset.download())
+
+    """
+
+    def __init__(self, credentials=None):
+        self._credentials = credentials or get_default_credentials()
+        self._context = context.create_context(self._credentials)
+
+    def geocode(self, dataset, street,
+                city=None, state=None, country=None,
+                metadata=None,
+                table_name=None, if_exists=Dataset.FAIL,
+                dry_run=False):
+        """Geocode a dataset
+
+        Args:
+            dataset (Dataset): a Dataset object to be geocoded.
+            street (str): name of the column containing postal addresses
+            city (str, optional): either the name of a column containing the addresses'
+                city names or a quoted literal value, e.g. "'New York'".
+            state (str, optional): either the name of a column containing the addresses'
+                State names or a quoted literal value, e.g. "'Illinois'".
+            country (str, optional): either the name of a column containing the addresses'
+                Country names or a quoted literal value, e.g. "'USA'".
+            metadata (str, optional): name of a column where metadata (in JSON format)
+                will be stored
+            table_name (str, optional): the geocoding results will be placed in a new
+                CARTO table with this name.
+            if_exists (str, optional): Behavior for creating new datasets, only applicable
+                if table_name isn't None;
+                Options are 'fail', 'replace', or 'append'. Defaults to 'fail'.
+            dry_run (bool, optional): no actual geocoding will be performed (useful to
+                check the needed quota)
+
+        Returns:
+            tuple: (Dataset, info_dict)
+
+        """
+
+        input_dataframe = None
+        if isinstance(dataset, pd.DataFrame):
+            input_dataframe = dataset
+            dataset = Dataset(input_dataframe)
+
+        if dry_run:
+            table_name = None
+
+        temporary_table = False
+
+        input_dataset = dataset
+        if input_dataset.is_remote() and input_dataset.table_name:  # FIXME: more robust to check first for query (hasattr(input_dataset, 'query'))
+            # input dataset is a table
+            if table_name:
```

Comment:
I feel this part should be solved by Dataset class and in some cases it could, like using `get_query` instead of doing it yourself. But I will go far on that: I have not checked it but I think you could do:
```
input_dataset.upload(input_table_name, ...)
```
in any case (dataset from table or query).